File: client/src/game/models/conversion/tracker.ts
import type { Tracker } from "../../shapes/interfaces";
import type { ServerTracker } from "../shapes";

export const trackersFromServer = (...trackers: ServerTracker[]): Tracker[] => {
    const result = [];
    for (const tracker of trackers) {
        result.push({
            uuid: tracker.uuid,
            name: tracker.name,
            visible: tracker.visible,
            value: tracker.value,
            maxvalue: tracker.maxvalue,
            draw: tracker.draw,
            primaryColor: tracker.primary_color,
            secondaryColor: tracker.secondary_color,
            temporary: false,
        });
    }
    return result;
};

export const trackersToServer = (shape: string, trackers: Tracker[]): ServerTracker[] => {
    const result = [];
    for (const tracker of trackers) {
        result.push({
            uuid: tracker.uuid,
            visible: tracker.visible,
            name: tracker.name,
            value: tracker.value,
            maxvalue: tracker.maxvalue,
            draw: tracker.draw,
            primary_color: tracker.primaryColor,
            secondary_color: tracker.secondaryColor,
            shape,
        });
    }
    return result;
};

export const partialTrackerToServer = (tracker: Partial<Tracker>): Partial<ServerTracker> => {
    return {
        uuid: tracker.uuid,
        visible: tracker.visible,
        name: tracker.name,
        value: tracker.value,
        maxvalue: tracker.maxvalue,
        draw: tracker.draw,
        primary_color: tracker.primaryColor,
        secondary_color: tracker.secondaryColor,
    };
};

export const partialTrackerFromServer = (tracker: Partial<ServerTracker>): Partial<Tracker> => {
    const partial: Partial<Tracker> = {};
    if ("uuid" in tracker) partial.uuid = tracker.uuid;
    if ("visible" in tracker) partial.visible = tracker.visible;
    if ("name" in tracker) partial.name = tracker.name;
    if ("value" in tracker) partial.value = tracker.value;
    if ("maxvalue" in tracker) partial.maxvalue = tracker.maxvalue;
    if ("draw" in tracker) partial.draw = tracker.draw;
    if ("primary_color" in tracker) partial.primaryColor = tracker.primary_color;
    if ("secondary_color" in tracker) partial.secondaryColor = tracker.secondary_color;
    return partial;
};

File: server/src/db/models/tracker.py
from __future__ import annotations

from typing import cast
from uuid import uuid4

from peewee import BooleanField, ForeignKeyField, IntegerField, TextField

from ...api.models.tracker import ApiTracker
from ..base import BaseDbModel
from .shape import Shape


class Tracker(BaseDbModel):
    uuid = cast(str, TextField(primary_key=True))
    shape = cast(Shape, ForeignKeyField(Shape, backref="trackers", on_delete="CASCADE"))
    visible = cast(bool, BooleanField())
    name = cast(str, TextField())
    value = cast(int, IntegerField())
    maxvalue = cast(int, IntegerField())
    draw = cast(bool, BooleanField())
    primary_color = cast(str, TextField())
    secondary_color = cast(str, TextField())

    def __repr__(self):
        return f"<Tracker {self.name} {self.shape.get_path()}>"

    def as_pydantic(self):
        return ApiTracker(
            uuid=self.uuid,
            shape=self.shape.uuid,
            visible=self.visible,
            name=self.name,
            value=self.value,
            maxvalue=self.maxvalue,
            draw=self.draw,
            primary_color=self.primary_color,
            secondary_color=self.secondary_color,
        )

    def make_copy(self, new_shape):
        _dict = self.as_pydantic()
        _dict.uuid = str(uuid4())
        _dict.shape = new_shape
        type(self).create(**_dict.model_dump())
